content/en/blog/running-gpu-based-functions-on-fission.md

```diff
@@ -0,0 +1,340 @@
+
+  ```bash
+  docker build -t USER/python-pytorch-env .
+  ```
+
+- Verify that image is created.
+
+  ```bash
+  $ docker images | grep python-pytorch-env
+  sohrab/python-pytorch-env                   latest            1032fa124b2b   2 days ago      20.5GB
+  ```
+
+#### Environment builder image
+
+- Replace the [Dockerfile](https://github.com/fission/environments/blob/master/python/Dockerfile) in [Python environments repository](https://github.com/fission/environments/tree/master/python) with following contents:
+
+  ```dockerfile
+  ARG BUILDER_IMAGE=fission/builder
+  ARG PLATFORM=linux/amd64
+
+  FROM ${BUILDER_IMAGE}
+  FROM --platform=${PLATFORM} nvcr.io/nvidia/pytorch:24.08-py3
+
+  COPY --from=0 /builder /builder
+  RUN apt-get update && apt-get install -y libev-dev libevdev2
+
+  ADD defaultBuildCmd /usr/local/bin/build
+
+  EXPOSE 8001
+  ```
+
+- Create `python-pytorch-builder` image using this updated Dockerfile.
+
+  ```bash
+  docker build -t USER/python-pytorch-builder .
+  ```
+
+- Verify that image is created.
+  
+  ```bash
+  $ docker images | grep python-pytorch-builder
+  USER/python-pytorch-builder               latest            3fa2801dcb1d   2 days ago      20.5GB
+  ```
+
+#### Push the images to a Container Registry
+
+- You can push the images to a container registry like GHCR or use them locally.
+
+  ```bash
+  docker push REGISTRY/USER/python-pytorch-env
+  docker push REGISTRY/USER/python-pytorch-builder
+  ```
+
+- Alternatively, you can also use the existing images which I have built and pushed to GHCR already.
+
+  ```bash
+  docker pull ghcr.io/soharab-ic/python-pytorch-env:latest
+  docker pull ghcr.io/soharab-ic/python-pytorch-builder:latest
+  ```
+
+### Verify the Environment with GPU
+
+In this step, we will do following things:
+
+- Create an environmnt in Fission using newly created environment and builder image.
+- Edit the environment deployment and add GPU resources and nodeSelector to make environment pod schedule on GPU nodes.
+- Create a function and verify the GPU availability inside the environment container.
+
+#### Fission Environment creation
+
+- Create Python environment using `python-pytorch-env` and `python-pytorch-builder` images.
+  
+  ```bash
+  fission env create --name python --image ghcr.io/soharab-ic/python-pytorch-env --builder ghcr.io/soharab-ic/python-pytorch-builder --poolsize 1
+  ```
+
+- The `fission env create` command will create two deployments. One deployment named `poolmgr-python-default-*` for environment and another for builder named `python-*`.
+- Patch the environment deployment to add GPU resources to `python` environment container and set `nodeSelector` to schedule pods on a GPU node using `kubectl patch` command.
+
+  ```bash
+  kubectl patch deployment poolmgr-python-default-5560759 -p '{"spec": {"template": {"spec":{"containers":[{"name":"python","resources": {"limits": {"nvidia.com/gpu": "1"}, "requests": {"nvidia.com/gpu": "1"}}}]}}}}'
```

Comment:
Please consider patching environment spec. Not deployment.
User should not change fissions internal object.

Reply:
@sanketsudake Patching the resources is working via environment spec but nodeSelector patch is not working.
NodeSelector field is ignored when creating deployment.
```
apiVersion: fission.io/v1
kind: Environment
metadata:
  creationTimestamp: null
  name: pytorch
spec:
  builder:
    command: build
    container:
      name: ""
      resources: {}
    image: ghcr.io/fission/python-builder
  imagepullsecret: ""
  keeparchive: false
  poolsize: 1
  resources:
    requests:
      nvidia.com/gpu: "1"
    limits:
      nvidia.com/gpu: "1"
  runtime:
    container:
      name: ""
      resources: {}
    image: ghcr.io/fission/python-env
    podSpec:
    nodeSelector:
      kubernetes.io/hostname: "gpu-node03"
  version: 3
```

Reply:
So typically you need to have container name same environment. As we have two containers in pod 
- runtime and fetcher 

containers is required field https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.26/#podspec-v1-core

https://github.com/fission/fission/blob/93869d3bc8b6097143e5488366dae178488b4474/test/tests/test_specs/test_spec_merge/specs/env-nodend.yaml#L25

Probably we can skip node selector part in blog post. only keep resources requests.

Reply:
Checkout 
- https://github.com/fission/fission/pull/2044
- https://github.com/fission/fission/issues/2043

Reply:
Skipping the node selector part.

Reply:
Yes please update blog accordingly